IBN-2078 add route to force hue bridge ip as discovery

Maw-Jung opened the pull request:
If a hue bridge isn't found through nupnp, the new route allows to manually
put a hue bridge in the discovery. This is particullary helpful when the
svs is offline.

Commit:
IBN-2078 add route to force hue bridge ip as discovery

If a hue bridge isn't found through nupnp, the new route allows to manually
put a hue bridge in the discovery. This is particullary helpful when the
svs is offline.

Commit:
IBN-2078 add discovery handler to access internal package

The upnp discovery package is internal and thus not available in ogsi.
Through an osgi known handler in the bundle the discovery can be accessed.

Commit:
Satisfy Code Analysis with missing license header

The GroupItemStateEvent had no license header.

mkhl commented on bundles/io/org.eclipse.smarthome.io.rest.core/META-INF/MANIFEST.MF:
```diff
@@ -17,13 +17,15 @@ Import-Package:
  com.google.common.base,
  com.google.common.collect,
  com.google.gson,
+ com.google.gson.reflect;version="2.7.0",
```
I’d leave of the version number since we always use the version defined in our upstream dependencies.

mkhl replied:
Also I don’t see a direct usage of this package, do we really need to import it?

mkhl commented on extensions/binding/org.eclipse.smarthome.binding.hue/OSGI-INF/HueBridgeNupnpDiscovery.xml:
```diff
@@ -17,5 +17,6 @@
    <implementation class="org.eclipse.smarthome.binding.hue.internal.discovery.HueBridgeNupnpDiscovery"/>
    <service>
       <provide interface="org.eclipse.smarthome.config.discovery.DiscoveryService"/>
+      <provide interface="org.eclipse.smarthome.binding.hue.internal.discovery.HueBridgeNupnpDiscovery"/>
```
I don’t see us referencing this interface, was that done outside of this pul request? Otherwise I don’t think we need to provide it. (Also providing an _internal_ interface like this seems wrong.)

Maw-Jung replied:
Right, i forgot to remove this, since the handler had the point to use the internal package.

mkhl commented on extensions/binding/org.eclipse.smarthome.binding.hue/OSGI-INF/HueBridgeDiscoveryHandler.xml:
```diff
@@ -0,0 +1,21 @@
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0" immediate="true" name="org.eclipse.smarthome.binding.hue.handler.hue.discovery.handler">
+   <implementation class="org.eclipse.smarthome.binding.hue.handler.HueBridgeDiscoveryHandler"/>
+   <service>
+      <provide interface="org.eclipse.smarthome.binding.hue.handler.HueBridgeDiscoveryHandler"/>
```
Huh, I didn’t think this would work when we pass a class name as the interface. TIL!

mkhl commented on extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/handler/HueBridgeDiscoveryHandler.java:
```diff
@@ -0,0 +1,34 @@
+     * @return DiscoveryResult
+     */
+    public DiscoveryResult addDiscovery(String ip) {
+        HueBridgeNupnpDiscovery discovery = new HueBridgeNupnpDiscovery();
```
`HueBridgeNupnpDiscovery` is declared as a Service, maybe we should reference (i.e. inject) it here instead of constructing a new instance?

mkhl replied:
Actually, if we inject a NupnpDiscovery, we could obviate this class entirely and just inject the NupnpDiscovery in the Resource instead…

Maw-Jung replied:
I wanted to stay away from injecting an internal package, thus the handler class.

mkhl commented on extensions/binding/org.eclipse.smarthome.binding.hue/src/main/java/org/eclipse/smarthome/binding/hue/internal/discovery/ManualBridgeJsonParameters.java:
```diff
@@ -0,0 +1,49 @@
+ */
+public class ManualBridgeJsonParameters {
+
+    @SerializedName("bridgeid")
```
Is there any reason not to use names consistent with BridgeJsonParameters, i.d. `id` instead of `bridgeId` and `macAddress` instead of `mac`?

Maw-Jung replied:
Streamlining the names could imply that the data is coming from the same source, this isn't the case tho, since the BridgeJsonParameters are red from a philips url and the ManualBridgeJsonParameters are red from a bridge directly. The serialization i'm doing is only changing the variables to camel case, so i'd like to stay away from renaming parameters.

mkhl replied:
I see, sounds good then.

mkhl commented:
retest this please

mkhl commented:
retest this please

mkhl commented:
retest this please

mkhl commented on bundles/io/org.eclipse.smarthome.io.rest.core/META-INF/MANIFEST.MF:
```diff
@@ -17,13 +17,15 @@ Import-Package:
  javax.ws.rs.ext,
  org.apache.commons.lang,
  org.eclipse.jdt.annotation;resolution:=optional,
+ org.eclipse.smarthome.binding.hue.handler,
```
After spending too much time with the error message, I believe this introduces a circular dependency between the OSGi package, which fails the feature verification because neither bundle can be installed without the other.

The Resource with the new Route now requires the org.eclipse.smarthome.binding.hue.handler package, but its package (org.eclipse.smarthome.io.rest.core) is part of the esh-base feature, which is required by pretty much every other feature, including the one containing the hue binding.

To fix the problem, I suggest creating a new Resource, either in the hue binding or in a new bundle altogether, that only provides the one route for forcibly adding a discovery result.
That would also fix the layering violation this changeset introduces in the resource, where before it handled all discovery services generically but now also knows of a specific discovery service and its specific additional capability, providing a route only for it.

An alternative way to go about this could be introducing another interface that extends DiscoveryService and adds a method for forcibly adding discovery results of some kind. The new route could include a binding id to help find the correct discovery service, then check that it implements the added interface, then use that to add the given manual “discovery result”.

Commit:
IBN-2078 Fix circular dependency in osgi

Through decoupling the generic discovery from the specific manual discovery with
introduction of a new resource, the cyclic dependency in osgi is resolved.
